### PR TITLE
Implemented hydra heads + adaptive kl

### PR DESCRIPTION
**Pull request by Dahoas**
Implemented BranchModel class to support multi-headed hydra type models. Also added adaptive kl controller.

Achieves 4x speedup for training on GPT2-mediuma and 10x speedup for training on GPTj and halves memory footprint.

I also added unittests.

### CODE REVIEW

File: trlx/model/nn/ppo_models.py [hunk 1]
```diff
@@ -100,3 +102,290 @@ def forward(
             cross_attentions=transformer_outputs.cross_attentions,
             value=value,
         )
+
+
+# Cell
+
+class ModelBranch(PreTrainedModel):
```
Comment (LouisCastricato) [inline trlx/model/nn/ppo_models.py:109]: Need a high level overview comment of how this class works.

File: unittests/test_ppo.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import unittest
```
Comment (LouisCastricato) [inline unittests/test_ppo.py:1]: Now this is a useful class but I think we should be handling unit in a separate PR....

Comment (Dahoas) [reply unittests/test_ppo.py:1]: I kept it in this merge because it tests the ModelBranch implementation

File: configs/ppo_config.yml [hunk 3]
```diff
@@ -37,13 +37,13 @@ method:
...
   init_kl_coef : 0.2  # init kl coefficient
-  target : 6  # target kl coefficient
+  target : 6  # target kl coefficient, set None for fixed kl coef
   horizon : 10000  # PPO horizon
   gamma : 1  # PPO discount
   lam : 0.95  # PPO lambda
   cliprange : 0.2  # clip range
   cliprange_value : 0.2  # clip range
-  vf_coef : 0.2  # value term weight
+  vf_coef : 2.3  # value term weight
```
[the remark this answers is not in the document]
Comment (Dahoas) [reply configs/ppo_config.yml:46]: I found these parameters work a lot better for quickly checking whether reward is increasing

File: configs/ppo_config.yml [hunk 4]
```diff
@@ -3,7 +3,7 @@ model:
   tokenizer_path : "gpt2"  # Name of hf tokenizer to load
   model_type : "AcceleratePPOModel"  # Name of accelerate model type to load
   device : "cuda"  # Train device
-  num_layers_unfrozen : -1  # Number of bottom layers to freeze during training
+  num_layers_unfrozen : 2  # Number of bottom layers to freeze during training
```
[the remark this answers is not in the document]
Comment (Dahoas) [reply configs/ppo_config.yml:6]: See comment below

File: unittests/test_ppo.py (hunk 2 above)
Comment (LouisCastricato) [reply unittests/test_ppo.py:1]: got it. It needs a lot of work. lets chat later.

File: configs/ppo_config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (LouisCastricato) [reply configs/ppo_config.yml:46]: Got it. sounds good.
